File: lib/weather/models/daily_weather.dart
import 'package:intl/intl.dart';

import 'weather_condition.dart';

class DailyWeather {
  final int epochTime;
  final String formattedTime;
  final String weekdayString;
  final DateTime sunrise;
  final DateTime sunset;
  final int maxTemperature;
  final int minTemperature;
  final WeatherCondition weatherCondition;
  final double pop;

  factory DailyWeather.fromJson(Map<String, dynamic> json) {

    try {
      // 시간 정보 - 1745982000
      final epochTime = json['dt'];

      // 시간 표시 - 오후 1시, 오후 3시 ...
      final DateTime dateTime = DateTime.fromMillisecondsSinceEpoch(
          epochTime * 1000, isUtc: true).toLocal();
      // 월/일 형태 (5/7)
      final String formattedTime = DateFormat('M/d').format(dateTime);
      // 요일 (월, 화, 수, ...)
      final String weekday = DateFormat.E('ko').format(dateTime);

      final DateTime sunrise = DateTime.fromMillisecondsSinceEpoch(json['sunrise'] * 1000, isUtc: true).toLocal();
      final DateTime sunset = DateTime.fromMillisecondsSinceEpoch(json['sunset'] * 1000, isUtc: true).toLocal();

      // 시간별 온도
      final int maxTemperature = json['temp']['max'].round();
      final int minTemperature = json['temp']['min'].round();

      // 시간별 날씨 상태 - 맑음, 비, 구름 ...
      final int weatherCode = json['weather'][0]['id'] as int;
      WeatherCondition weatherCondition = WeatherConditionX.fromWeatherCode(
          weatherCode);

      final double pop = (json['pop'] as num).toDouble();

      return DailyWeather(
        epochTime: epochTime,
        formattedTime: formattedTime,
        weekdayString: weekday,
        sunrise: sunrise,
        sunset: sunset,
        maxTemperature: maxTemperature,
        minTemperature: minTemperature,
        weatherCondition: weatherCondition,
        pop: pop,
      );
    } catch (e, stacktrace) {
      // 에러 로그 출력
      print('DailyWeather.fromJson 에러: $e');
      print(stacktrace);

      return DailyWeather.empty;
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'dt': epochTime,
      'sunrise': sunrise.toUtc().millisecondsSinceEpoch ~/ 1000,
      'sunset': sunset.toUtc().millisecondsSinceEpoch ~/ 1000,
      'temp': {
        'max': maxTemperature,
        'min': minTemperature,
      },
      'weather': [
        {
          'id': weatherCondition.toWeatherCode(),
        }
      ],
      'pop': pop,
    };
  }

  DailyWeather({
    required this.epochTime,
    required this.formattedTime,
    required this.weekdayString,
    required this.sunrise,
    required this.sunset,
    required this.maxTemperature,
    required this.minTemperature,
    required this.weatherCondition,
    required this.pop
  });

  static final empty = DailyWeather(
    epochTime: 1746846000,
    formattedTime: '5/11',
    weekdayString: '일',
    sunrise: DateTime.fromMillisecondsSinceEpoch(1746822585 * 1000, isUtc: true).toLocal(),
    sunset: DateTime.fromMillisecondsSinceEpoch(1746872828 * 1000, isUtc: true).toLocal(),
    maxTemperature: 0,
    minTemperature: 0,
    weatherCondition: WeatherCondition.unknown,
    pop: 0.0,
  );
}

File: ios/Runner/Models/DailyWeather.swift
//
//  DailyWeather.swift
//  Runner
//
//  Created by 김두원 on 5/23/25.
//

import Foundation

struct DailyWeather {
    var epochTime: Date
    var formattedTime: String
    var weekdayString: String
    var sunrise: Date
    var sunset: Date
    var maxTemperature: Int
    var minTemperature: Int
    var weatherCondition: WeatherCondition
    var pop: Double
    
    static func fromEntity(entity: DailyWeatherEntity) -> DailyWeather {
        let timeInterval = TimeInterval(entity.dt)
        let epochTime = Date(timeIntervalSince1970: timeInterval)
        
        let dateFormatter1 = DateFormatter()
        dateFormatter1.locale = Locale(identifier: "ko_KR")
        dateFormatter1.dateFormat = "M/d"

        let dateFormatter2 = DateFormatter()
        dateFormatter2.locale = Locale(identifier: "ko_KR")
        dateFormatter2.dateFormat = "E"  // 요일(월, 화, 수..)

        let formattedTime = dateFormatter1.string(from: epochTime)
        let weekdayString = dateFormatter2.string(from: epochTime)
        
        let sunrise = Date(timeIntervalSince1970: TimeInterval(entity.sunrise))
        let sunset = Date(timeIntervalSince1970: TimeInterval(entity.sunset))
        
        let minTemperature = Int(entity.temp.min.rounded())
        let maxTemperature = Int(entity.temp.max.rounded())
        
        let weatherCode = entity.weather.first?.id ?? 0
        let weatherCondition = WeatherCondition.fromWeatherCode(code: weatherCode)
        
        let pop = entity.pop

        return DailyWeather(
            epochTime: epochTime,
            formattedTime: formattedTime,
            weekdayString: weekdayString,
            sunrise: sunrise,
            sunset: sunset,
            maxTemperature: maxTemperature,
            minTemperature: minTemperature,
            weatherCondition: weatherCondition,
            pop: pop
        )
    }
    
    static let dummy = DailyWeather.fromEntity(entity: DailyWeatherEntity.dummy)
}

struct DailyWeatherEntity: Codable {
    let dt: Int
    let sunrise: Int
    let sunset: Int
    let moonrise: Int
    let moonset: Int
    let moonPhase: Double
    let summary: String
    let temp: Temperature
    let feelsLike: FeelsLike
    let pressure: Int
    let humidity: Int
    let dewPoint: Double
    let windSpeed: Double
    let windDeg: Int
    let windGust: Double
    let weather: [WeatherInfoEntity]
    let clouds: Int
    let pop: Double
    let rain: Double?
    let uvi: Double

    enum CodingKeys: String, CodingKey {
        case dt, sunrise, sunset, moonrise, moonset
        case moonPhase = "moon_phase"
        case summary, temp
        case feelsLike = "feels_like"
        case pressure, humidity
        case dewPoint = "dew_point"
        case windSpeed = "wind_speed"
        case windDeg = "wind_deg"
        case windGust = "wind_gust"
        case weather, clouds, pop, rain, uvi
    }

    struct Temperature: Codable {
        let day: Double
        let min: Double
        let max: Double
        let night: Double
        let eve: Double
        let morn: Double
    }

    struct FeelsLike: Codable {
        let day: Double
        let night: Double
        let eve: Double
        let morn: Double
    }
    
    static let dummy = DailyWeatherEntity(
        dt: 1747969200,
        sunrise: 1747945192,
        sunset: 1747996674,
        moonrise: 1747934940,
        moonset: 1747980480,
        moonPhase: 0.84,
        summary: "Expect a day of partly cloudy with rain",
        temp: DailyWeatherEntity.Temperature(
            day: 18.9,
            min: 16.81,
            max: 21.58,
            night: 17.2,
            eve: 20.34,
            morn: 16.86
        ),
        feelsLike: DailyWeatherEntity.FeelsLike(
            day: 18.26,
            night: 16.7,
            eve: 20.07,
            morn: 16.66
        ),
        pressure: 1011,
        humidity: 54,
        dewPoint: 10.42,
        windSpeed: 4.85,
        windDeg: 100,
        windGust: 7.69,
        weather: [
            WeatherInfoEntity(
                id: 500,
                main: "Rain",
                description: "실 비",
                icon: "10d"
            )
        ],
        clouds: 100,
        pop: 0.94,
        rain: 1.4,
        uvi: 6.86
    )
}
